### PR TITLE
refactoring so async, context manager and coroutine can live together

### PR DESCRIPTION
**Pull request by Bgouveia**
Description: This PR adds circuit breaker as a decorator and refactors the structure so we can provide this mechanism as a coroutine, async/await and as a context manager with proper inheritance.

ABOUTS THE TESTS: I was not able to properly configure the test for coroutines without. However, there is a test that validates the overall mechanics for this implementation: https://github.com/luizalabs/acme/pull/518

 ### Checklist:
 - [x] Tests
 - [x] Changelog
 - [ ] Docs (WIP)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -39,10 +39,8 @@ pip-delete-this-directory.txt
 # Unit test / coverage reports
 htmlcov/
 .tox/
-.coverage
```
Comment (luizgois-luizalabs) [inline .gitignore:42]: Algum motivo para remove do `.gitignore`?

File: asyncio_toolkit/circuit_breaker/async_await.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            self.circuit_timeout
+        )
+
+    def __cocall__(self, method):
+        @wraps(method)
+        async def wrapper(obj, *args, **kwargs):
+            await self.is_circuit_open()
+
+            try:
+                val = await method(obj, *args, **kwargs)
```
Comment (luizgois-luizalabs) [inline asyncio_toolkit/circuit_breaker/async_await.py]: O que acha de mudar de `val` para `response`?

File: .gitignore (hunk 1 above)
Comment (Bgouveia) [reply .gitignore:42]: @luizgois-luizalabs por causa do report do plugin q estamos usando no travis, ele precisa que exista uma versão comitada do report pra fazer o diff: https://docs.codecov.io/docs/error-reference#section-missing-base-commit

no PR anterior existia um erro por causa disso: https://github.com/luizalabs/asyncio-toolkit/pull/3

**Review by vtrmantovani — APPROVED** [review]

File: asyncio_toolkit/circuit_breaker/async_await.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+import asyncio
+import logging
+from functools import wraps
+
+from .base import BaseCircuitBreaker
+
+logger = logging.getLogger(__name__)
+
+
+class circuit_breaker(BaseCircuitBreaker):
```
Comment (henriquebraga-luizalabs) [inline asyncio_toolkit/circuit_breaker/async_await.py:9]: Out of curiosity: why class name is not CamelCase (`CircuitBreaker`)?

Comment (henriquebraga-luizalabs) [reply asyncio_toolkit/circuit_breaker/async_await.py:9]: Oh, I see. It's a decorator 😄

File: asyncio_toolkit/circuit_breaker/base.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        This method demands that the implementation is responsible for
+        getting a storage key from the storage engine.
+        """
+
+    @property
+    @abc.abstractmethod
+    def is_circuit_open(self):
+        """
+        This method must implement a call to the storage engine in order
+        to retrieve the current circuit state, based on this intance's key.
```
Comment (nicolasalves-luizalabs) [inline asyncio_toolkit/circuit_breaker/base.py]: `intances`

**Review by nicolasalves-luizalabs — APPROVED** [review]

**Review by luizgois-luizalabs — APPROVED** [review]

File: asyncio_toolkit/circuit_breaker/base.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    @abc.abstractmethod
+    def open_circuit(self):
+        """
+        This method must "open the circuit" by setting this instance's key
+        on the storage engine, and also, setting a TTL to it.
+        """
+
+    def _is_catchable(self, exception):
+        logger.debug('Testing if {} is catcheable'.format(type(exception)))
+        return any(
```
Comment (henriquebraga-luizalabs) [inline asyncio_toolkit/circuit_breaker/base.py]: What do you think to say here if the exception is indeed catchable?
```python
is_catchable = any(exc in self.catch_exceptions
                               for exc in [type(exception), exception])

logger.debug('Testing if {} is catchable: {}'.format(type(exception), is_catchable))
```

**Review by henriquebraga-luizalabs — APPROVED** [review]

**Pull request merged**